pkg/bloomcompactor/chunkcompactor.go

```diff
@@ -0,0 +1,184 @@
+package bloomcompactor
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/pkg/logproto"
+	v1 "github.com/grafana/loki/pkg/storage/bloom/v1"
+	"github.com/grafana/loki/pkg/storage/bloom/v1/filter"
+	"github.com/grafana/loki/pkg/storage/chunk"
+	"github.com/grafana/loki/pkg/storage/stores/shipper/bloomshipper"
+	tsdbindex "github.com/grafana/loki/pkg/storage/stores/shipper/indexshipper/tsdb/index"
+)
+
+type compactorTokenizer interface {
+	PopulateSeriesWithBloom(bloom *v1.SeriesWithBloom, chunks []chunk.Chunk)
+}
+
+type chunkClient interface {
+	// TODO: Consider using lazyChunks to avoid downloading all requested chunks.
+	GetChunks(ctx context.Context, chunks []chunk.Chunk) ([]chunk.Chunk, error)
+}
+
+type blockBuilder interface {
+	BuildFrom(itr v1.Iterator[v1.SeriesWithBloom]) (uint32, error)
+	Data() (io.ReadCloser, error)
+}
+
+type PersistentBlockBuilder struct {
+	builder  *v1.BlockBuilder
+	localDst string
+}
+
+func NewPersistentBlockBuilder(localDst string, blockOptions v1.BlockOptions) (*PersistentBlockBuilder, error) {
+	// write bloom to a local dir
+	b, err := v1.NewBlockBuilder(blockOptions, v1.NewDirectoryBlockWriter(localDst))
+	if err != nil {
+		return nil, err
+	}
+	builder := PersistentBlockBuilder{
+		builder:  b,
+		localDst: localDst,
+	}
+	return &builder, nil
+}
+
+func (p *PersistentBlockBuilder) BuildFrom(itr v1.Iterator[v1.SeriesWithBloom]) (uint32, error) {
+	return p.builder.BuildFrom(itr)
+}
+
+func (p *PersistentBlockBuilder) Data() (io.ReadCloser, error) {
+	blockFile, err := os.Open(filepath.Join(p.localDst, v1.BloomFileName))
```

Comment:
bloomClient will close the reader once the uploading is completed, but we also need to remove the file from the local disk. Do we have such mechanisms?